### PR TITLE
Feature/billionaire

### PR DESCRIPTION
**Pull request by ngevangelion**
megamillion api and data binding to page

### CODE REVIEW

**Comment by ngevangelion** [comment]
@lufli 
got this issue while doing yarn install/add
error An unexpected error occurred: "EPERM: operation not permitted, unlink 'C:\\Users\\chris\\next-tool\\node_modules\\@next\\swc-win32-x64-msvc\\next-swc.win32-x64-msvc.node'".

causing yarn-lock file not updated and build failed any idea how to resolve?

File: lib/lottery/megamillion.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      break;
+    }
+  }
+  if (end < 0) {
+    return '';
+  }
+  return rawData.substring(start, end + 1);
+};
+
+export async function getWinningMegaNumber() {
```
Comment (lufli) [inline lib/lottery/megamillion.js]: I prefer function name `getWinningNumbers`, and it should return an object like this:
``` json
{
  type: 'mega',
  date: '2023-07-14T17:45:41.487Z', // new Date(),
  whiteBalls: [1, 2, 3, 4, 5], // or make one property for one number: N1, N2, N3, N4, N5, and NS
  specialBall: 6,
  multiplier: 2,
}
```

And for function `checkNumbersForPrize`, return an Check like this:
```

  Check {
    type: 'powerball',
    date: '2023-07-14T17:45:41.487Z',
    whiteBalls: [
      {
        number: 1,
        hit: true,
      },
      {
        number: 2,
        hit: true,
      },
    ],
    specialBall: {
      number: 6,
      hit: false,
    },
    multiplier: 2,
    prize: 14,
  }
```

File: lib/lottery/megamillion.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+function checkNumbersForPrize(winningNumber, playNumber) {
+  let numberOfWhiteBallHits = 0;
+  let winningResult = {
+    type: 'megamillion',
+    date: new Date(),
+    multiplier: winningNumber.multiplier
+  };
+
+  for (let i = 1; i < 6; i++) {
+    let hit = Number(playNumber[`N${i}`]) === Number(winningNumber[`N${i}`]);
```
Comment (lufli) [inline lib/lottery/megamillion.js]: should check whether or not all white ball numbers includes here. not first ball === first play, second ball === second play, and so on...

File: lib/lottery/powerball.js [hunk 3]
```diff
@@ -31,7 +31,7 @@ async function getWinningNumbers() {
   const multiplier = parseInt(dom.getElementsByClassName('multiplier')[0].innerHTML, 10);
   return {
     type: 'powerball',
-    date: new Date(),
+    date: Date.now(),
```
Comment (lufli) [inline lib/lottery/powerball.js:34]: Good catch! 👍 please help fix line 134, and double-play.js.

File: lib/lottery/megamillion.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  return response.data;
+}
+
+function parseRawData(rawData) {
+  let start = rawData.indexOf('{');
+  let end;
+  if (start < 0) {
+    return '';
+  }
+  for (let i = rawData.length -1; i > 0; i--) {
```
Comment (lufli) [inline lib/lottery/megamillion.js]: I found there is a function we can use here: String.prototype.lastIndexOf()

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/lastIndexOf

File: lib/lottery/megamillion.js (hunk 2 above)
Comment (ngevangelion) [reply lib/lottery/megamillion.js]: fixed, please check

File: lib/lottery/megamillion.js (hunk 4 above)
Comment (ngevangelion) [reply lib/lottery/megamillion.js]: added

File: lib/lottery/double-play.js [hunk 5]
```diff
@@ -29,7 +29,7 @@ async function getWinningNumbers() {
   const NS = parseInt(dom.getElementsByClassName('dp-powerball')[0].innerHTML, 10);
   return {
     type: 'double-play',
-    date: new Date(),
+    date: Date.now(),
```
Comment (lufli) [inline lib/lottery/double-play.js:32]: need to fix line 114 as well

File: lib/lottery/megamillion.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        return 0;
+    }
+  }
+}
+
+function checkNumbersForPrize(winningNumber, playNumber) {
+  let numberOfWhiteBallHits = 0;
+  const winningResult = {
+    type: 'megamillions',
+    date: new Date(),
```
Comment (lufli) [inline lib/lottery/megamillion.js]: we should use Date.now() for timestamps

File: lib/lottery/megamillion.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    N2,
+    N3,
+    N4,
+    N5,
+  };
+
+  finalResult.NS = MBall;
+  finalResult.multiplier = Number(Megaplier);
+  finalResult.date = PlayDate;
+  finalResult.jackpot = dataJson.Jackpot;
```
Comment (lufli) [inline lib/lottery/megamillion.js:39]: Dont use playdate here, the timestamp here is the time we generate the result.
why we need jackpot?

File: lib/lottery/megamillion.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+  const winningWhiteBalls = [
+    Number(winningNumber.N1),
+    Number(winningNumber.N2),
+    Number(winningNumber.N3),
+    Number(winningNumber.N4),
+    Number(winningNumber.N5),
+  ];
+
+  for (let i = 1; i < 6; i++) {
```
Comment (lufli) [inline lib/lottery/megamillion.js:110]: We would better use higher order functions here. But it is ok for now.

File: lib/lottery/megamillion.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    winningResult[`N${i}`] = {
+      hit,
+      number: Number(playNumber[`N${i}`]),
+    };
+
+    numberOfWhiteBallHits += Number(hit);
+  }
+
+  const specialHit = Number(winningNumber.NS) === Number(playNumber.NS);
+  winningResult.NS = {
```
Comment (lufli) [inline lib/lottery/megamillion.js:121]: Need Date.now() here.

File: pages/billionaire.js [hunk 10]
```diff
@@ -1,7 +1,16 @@
 import Head from 'next/head';
+import axios from 'axios';
 import React from 'react';
+import useSwr from 'swr';
+import ResultRender from '@/components/billionaire/ResultRender';
 
 export default function Billionaire() {
+  const { data, mutate } = useSwr('/api/get-lottery-results', axios.get, {
```
Comment (lufli) [inline pages/billionaire.js:8]: hahaha... You dont need to implement this page now. But it is ok leave it there.

File: package.json [hunk 11]
```diff
@@ -30,6 +32,8 @@
     "eslint-plugin-node": "^11.1.0",
     "eslint-plugin-react": "^7.28.0",
     "eslint-plugin-react-hooks": "^4.3.0",
+    "fetch-mock": "^9.11.0",
+    "isomorphic-fetch": "^3.0.0",
```
Comment (lufli) [inline package.json]: do we need  "fetch-mock" and "isomorphic-fetch"?

**Review by lufli — COMMENTED** [review]
牛逼 👍 ! But leave several small things to fix.

File: lib/lottery/megamillion.js (hunk 7 above)
Comment (ngevangelion) [reply lib/lottery/megamillion.js:39]: show how much we can earn!

**Review by lufli — APPROVED** [review]
牛逼

**Pull request merged**